modules/gapi/test/serialization/serialization.cpp

```diff
@@ -0,0 +1,1292 @@
+    }
+    return os;
+}
+
+SerializationStream&  operator << (SerializationStream& os, double double_val)
+{
+    uint element_tmp[2];
+    memcpy(&element_tmp, &double_val, 2 * sizeof(uint));
+    os << element_tmp[0];
+    os << element_tmp[1];
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::vector<cv::Size> cvsizes)
+{
+    os << (uint)cvsizes.size();
+    for (auto & element : cvsizes)
+    {
+        os << element;
+    }
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, cv::Size cvsize)
+{
+    os << cvsize.width;
+    os << cvsize.height;
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::vector<bool> bools)
+{
+    uint bools_size = (uint)bools.size();
+    os << bools_size;
+    for (uint j = 0; j < bools_size; j++)
+    {
+        int bool_val_int;
+        bool_val_int = bools[j] ? 1 : 0;
+        os << bool_val_int;
+        //os << bools[j];
+    }
+    return os;
+}
+
+SerializationStream&  operator << (SerializationStream& os, bool bool_val)
+{
+    int bool_val_int;
+    bool_val_int = bool_val ? 1 : 0;
+    os << bool_val_int;
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::vector<cv::Scalar> cvscalars)
+{
+    os << (uint)cvscalars.size();
+    for (auto & element : cvscalars)
+    {
+       os << element;
+    }
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, cv::Scalar cvscalar)
+{
+    uint element_tmp[2];
+    for (uint i = 0; i < 4; i++)
+    {
+        memcpy(&element_tmp, &cvscalar.val[i], 2 * sizeof(uint));
+        os << element_tmp[0];
+        os << element_tmp[1];
+    }
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::vector<cv::Point> cvpoints)
+{
+    os << (uint)cvpoints.size();
+    for (auto & element : cvpoints)
+    {
+        os << element;
+    }
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, cv::Point cvpoint)
+{
+    os << cvpoint.x;
+    os << cvpoint.y;
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::vector<cv::Mat> cvmats)
+{
+    os << (uint)cvmats.size();
+    for (auto & element : cvmats)
+    {
+        os << element;
+    }
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, cv::Mat cvmat)
+{
+    size_t matSizeInBytes = cvmat.rows * cvmat.step[0];
+    int mat_type = cvmat.type();
+    os << cvmat.cols;
+    os << cvmat.rows;
+    os << mat_type;
+    os << (uint)cvmat.step[0];
+
+    if (matSizeInBytes != 0)
+    {
+        size_t numAtoms = matSizeInBytes % sizeof(uint) ==
+            0 ? (matSizeInBytes / sizeof(uint)) : (matSizeInBytes / sizeof(uint)) + 1;
+        uint* hton_buff = (uint*)malloc(numAtoms * sizeof(uint));
+        memcpy(hton_buff, cvmat.data, matSizeInBytes);
+        for (uint a = 0; a < numAtoms; a++)
+        {
+            os << hton_buff[a];
+        }
```

Comment:
no, please don't do this. Use a bulk write. let "char*" be an "atom" (given you know how many bytes to read and write).

What if the Mat is an ROI? You'll read an incorrect region (discarding the stride)